kanidmd/lib/src/entry.rs

```diff
@@ -1795,6 +1795,24 @@ impl Entry<EntryReduced, EntryCommitted> {
                         atype: "entrydn".to_string(),
                         vals: vec![dn.as_bytes().to_vec()],
                     }),
+                    "mail;primary" | "mailprimary" | "emailprimary" => {
+                        attr_map.get(kani_a).map(|pvs| LdapPartialAttribute {
+                            atype: ldap_a.to_string(),
+                            vals: pvs
```

Comment:
I think this here isn't right. The pvs is a `Vec<Vec<u8>>` which contains a set of emails per https://github.com/kanidm/kanidm/blob/master/kanidmd/lib/src/valueset/address.rs#L321 but this isn't sorted so that the primary is first, it's just the set of addresses with the primary tagging removed. 

I think to make this "consistent" with your behaviour, you need to adjust to_proto_string_clone_iter in address.rs to put the primary first so that your .first() and .split_first() trick works correctly. 

This probably explains the issue you were seeing with the incorrect order of primary based on what you set on the cli. 

It feels a bit like a hack to do it this way and rely on the ordering, so I want to think about this some more for a bit too, but for now that will at least make this work.

Additionally, please add a test for this in kanidmd/lib/src/ldap.rs to ensure that the virtual attributes are being generated as we expect so that we don't regress or break this in future!

Reply:
Thanks for the hint. I added that feature. and also implemented tests.